
# HG changeset patch
# User Todd C. Miller <Todd.Miller@courtesan.com>
# Date 1389722096 25200
# Node ID a204b69d91f786444899ce3fc95ed535b6e32d33
# Parent  987087ce465828da121b9d67412f42feaea06d56
If not logging I/O we may get EOF when the command is executed and
the other end of the backchannel is closed.  Just remove the
backchannel event in this case or we will continue to receive
the event.  Bug #631

diff -r 987087ce4658 -r a204b69d91f7 src/exec.c
--- a/src/exec.c	Tue Jan 14 09:26:14 2014 -0700
+++ b/src/exec.c	Tue Jan 14 10:54:56 2014 -0700
@@ -230,11 +230,14 @@
 	    /* Short read or EOF. */
 	    sudo_debug_printf(SUDO_DEBUG_ERROR,
 		"failed to read child status: %s", n ? "short read" : "EOF");
-	    /*
-	     * If not logging I/O we may get EOF when the command is
-	     * executed and sv is closed.  It is safe to ignore this.
-	     */
-	    if (ec->log_io || n != 0) {
+	    if (!ec->log_io && n == 0) {
+		/*
+		 * If not logging I/O we may get EOF when the command is
+		 * executed and the other end of the backchannel is closed.
+		 * Just remove the event in this case.
+		 */
+		(void)sudo_ev_del(ec->evbase, backchannel_event);
+	    } else {
 		/* XXX - need new CMD_ type for monitor errors. */
 		errno = n ? EIO : ECONNRESET;
 		ec->cstat->type = CMD_ERRNO;

